### ISSUE TITLE
Discussion: TypeScript for development?

### ISSUE THREAD

**Issue by lars-reimann** [Working-Title-MSFS-Mods/fspackages#509]
If I understand it correctly, the JavaScript code we have available right now was originally compiled TypeScript from Asobo. At least it was compiled _somehow_, based on the `//# sourceMappingURL=BaseNDCompass.js.map` at the bottom of most files.

Would there be any interest in moving our codebase back to TypeScript? Since TypeScript is a superset of JavaScript, we would not need to do this in some sort of big bang but could gradually make changes and improvements as we go along.

By doing this we could profit from static type checking and generally better IDE support for TypeScript, which should lead to more robust and faster development.

Some wishful thinking: Given that Asobo was mentioning this team in their latest Q&A and wanted to work with us, there might be a slight chance to get the TypeScript source files from them. That said, even if we get them, the question remains whether it would be easier to change our current state to TypeScript or to integrate the changes that already happened here into Asobo's code. The original TypeScript code could still be interesting as a reference, though.

**Comment by knighty** [comment]
Yeah we have been considering it from the start. The annoying part is that we have sort of 2 levels of dependencies here, in that some of the code we modify is dependent on asobo code and vice versa, making a project for it more complicated than it'd otherwise be. Access to the source files may make this easier for sure, but yes, we've certainly been considering it ;)

**Comment by lars-reimann** [comment]
Another reason that led me to believe that TypeScript was used to some extent was how some objects look if you inspect them in the Chrome console using [this](https://github.com/dga711/msfs-webui-devkit) and [this](https://github.com/dga711/devtools-backend-refurb). There you find that the `AutopilotMode` object looks like this:
```
{
    "MANAGED": 0,
    "SELECTED": 1,
    "HOLD": 2,
    0: "MANAGED",
    1: "SELECTED",
    2: "HOLD"
}
```

This is exactly the structure you get when you compile an enum in TypeScript:

```
// TypeScript
enum AutopilotMode {
    MANAGED,
    SELECTED,
    HOLD
}
```

```
// Compiled JavaScript
AutopilotMode[AutopilotMode["MANAGED"] = 0] = "MANAGED";
AutopilotMode[AutopilotMode["SELECTED"] = 1] = "SELECTED";
AutopilotMode[AutopilotMode["HOLD"] = 2] = "HOLD";
```

**Comment by dga711** [comment]
Yeah these ones are also the ones that led me to believes its compiled from TS, cause its bonkers

**Comment by mattnischan** [comment]
Yeah, it definitely looks exactly like current TS output for sure. The project structure is a little odd, normally you don't just blast out straight bunches of js files from ts, but it's definitely very possible to do so.

That being said, we would want to do this by attrition. The codebase at current already navigates relatively well with reference files in the right place, and so just converting everything we have to TS is not only a pretty big task but doesn't necessarily provide a ton of value for the time commitment. Obviously, we can rename all the js files to .ts, but that doesn't really help us enormously.

However, going forward with fully custom stuff, my opinion is that TS would be a better choice overall, given the size of the codebase.

**Comment by lars-reimann** [comment]
>However, going forward with fully custom stuff, my opinion is that TS would be a better choice overall, given the size of the codebase.

Yes, that is pretty much my take on it as well. For the existing code some minor changes like replacing `==` with `===`, `var` with `let`/`const` etc. that can be automatically detected (and fixed) to follow best practices would be recommendable in my opinion. We'd just have to be mindful of merge conflicts. 

Anything beyond that is optional, although I'd personally still encourage (but certainly not require) that whenever large changes are made to, say, an existing function the containing file is changed to TS and the function gets type annotations and documentation.
